.travis.yml

```diff
@@ -19,21 +19,20 @@ install:
  - ./gradlew flywayMigrate
 before_script: ./.travis/setup_gpg
 script:
-- ./gradlew --parallel check jacocoTestReport
+- ./gradlew --parallel -PproductVersionQualifier="$TRAVIS_BUILD_NUMBER" check jacocoTestReport
```

Comment:
The build number was passed to gradle previously in just one place, is it really needed now in both gradle script invocations? For a moment we were going to have the build number in travis env variable 100%, and the build file was going to know nothing at all about the game engine version number. So I'm partly concerned that the build script is becoming more and more aware of the game's concept of version number.

Reply:
> The build number was passed to gradle previously in just one place, is it really needed now in both gradle script invocations?

Yes, it's required in both places because of the Gradle incremental build process that does not re-run tasks whose inputs have not changed since the previous run.

On the first invocation, the `:game-core:processResources` task is run and will replace the `@qualifier@` token in _product.properties_ and copy the output to the "compiled state" of the project under the _game-core/build_ folder.  On the second invocation, the inputs to `:game-core:processResources` have not changed, and the task will not be re-run.  Thus, Gradle simply collects the compiled classes and resources from under _game-core/build_ and bundles them up into the artifacts.  So, if the build number was not passed during the first invocation, the `@qualifier@` token would end up in the final build artifact generated during the second invocation.

The build number is required on the second invocation because that's when the artifacts are produced.  Since the artifact names are based on the `version` project property, it must be correctly initialized during this invocation.

This is only an issue because the artifacts are only generated for builds on the `master` branch.  If we always ran the `release` task, this smell would go away.  But the downside of that approach is that PR builds will take a few minutes longer to generate the artifacts.  On the other hand, if we always ran the `release` task we wouldn't have the issue where a PR build can pass, but the `master` build subsequently fails after merge because part of the build script didn't get run on the PR branch.

Another alternative is to write this property to _~/.gradle/gradle.properties_ at the start of the build so it is picked up by all Gradle invocations now or in the future.  The only drawback there is that the data dependency is not explicitly visible within _.travis.yml_.

> So I'm partly concerned that the build script is becoming more and more aware of the game's concept of version number.

If by "build script" you mean _.travis.yml_, I'm not sure I agree.  It previously had even more intimate knowledge of the game's version number because it was responsible for defining it (i.e. `ENGINE_VERSION="1.9.0.0.${TRAVIS_BUILD_NUMBER}"`).  Now, _.travis.yml_ only knows about the qualifier (which is under its control) and how to ask Gradle for the full version number (to use for Git tagging).  It knows nothing about the semantic portion (e.g. `1.9.0.0`).

If you mean the Gradle build script, I'd agree there's a very slight increase in knowledge.  It already knew about the semantic portion because it had to read _game_engine.properties_.  Now, the additional responsibility is that it simply has to combine the semantic portion with the qualifier to generate the full version.  But, IMO, that increase is offset by the decrease of knowledge in _.travis.yml_.

Reply:
> Another alternative is to write this property to _~/.gradle/gradle.properties_ at the start of the build so it is picked up by all Gradle invocations now or in the future. The only drawback there is that the data dependency is not explicitly visible within _.travis.yml_.

00cfb5e03 implements this idea.  It does clean up _.travis.yml_ considerably, but, as I said above, at the cost of some data dependency visibility.  I can revert if you prefer the original.